kaggle-riiid/preprocessing/refinement.py

```diff
@@ -0,0 +1,57 @@
+import pandas as pd
+import random
+
+def refine_log_data(data_frame: pd.DataFrame,
+                    random_seed: int = 3141592,
+                    upper_bound: int = 10) -> tuple:
+    """Refine the log data to construct cross-sectional data.
+
+    Parameter
+    ========
+    `data_frame`: The content spend history log data.
+
+    `random_seed`: Control the random seed to cut the log data.
+    
+    `upper_bound`: The function will cut at least `upper_bound`th log data for each user.
+    """
+    random.seed(random_seed)
+
+    log_count_data = derive_random_cutoff_data(data_frame, upper_bound)
+    
+    log_data = pd.merge(data_frame, log_count_data, on = 'user_id')
+    log_data['row_id_by_user'] = log_data.groupby(['user_id'])['row_id'].rank()
+
+    train_log_data  = log_data[log_data['row_id_by_user'] < log_data['cutoff_position']]
+    test_log_data = log_data[log_data['row_id_by_user'] == log_data['cutoff_position']]
+
```

Comment:
You said that we will randomly select a upper bound user-wisely, but there is no part of random upper bound.

Reply:
I mistakes denoting the term of upper bound. In fact, the upper_bound variable is the same mean as lower bound.
I think the description for this branch is ambiguous. My original intention is that the cutoff position is random, and the setting the lower_bound is to regularize the random value so that the cut-offed log are sufficiently large.